Add mandatory OpenBSD bits

mptre opened the pull request:
This is the largest PR related to adding OpenBSD support but hard to split up.

dvyukov commented on pkg/report/report.go:
```diff
@@ -78,6 +78,7 @@ var ctors = map[string]fn{
 	"gvisor":  ctorGvisor,
 	"freebsd": ctorFreebsd,
 	"netbsd":  ctorNetbsd,
+	"openbsd": ctorOpenbsd,
```
This should go into the other commit.

mptre replied:
Fixed

dvyukov commented:
These test failures on bot look real:
```
        		<stdin>:92:2: error: unknown type name ‘event_t’
        		<stdin>: In function ‘thr’:
        		<stdin>:103:3: error: implicit declaration of function ‘event_wait’ [-Werror=implicit-function-declaration]
        		<stdin>:104:3: error: implicit declaration of function ‘event_reset’ [-Werror=implicit-function-declaration]
        		<stdin>:107:3: error: implicit declaration of function ‘event_set’ [-Werror=implicit-function-declaration]
        		<stdin>: In function ‘execute_one’:
        		<stdin>:122:5: error: implicit declaration of function ‘event_init’ [-Werror=implicit-function-declaration]
        		<stdin>:127:4: error: implicit declaration of function ‘event_isset’ [-Werror=implicit-function-declaration]
        		<stdin>:135:4: error: implicit declaration of function ‘event_timedwait’ [-Werror=implicit-function-declaration]
        		<stdin>: In function ‘loop’:
        		<stdin>:179:3: error: implicit declaration of function ‘remove_dir’ [-Werror=implicit-function-declaration]
        		<stdin>: In function ‘execute_call’:
        		<stdin>:190:3: error: implicit declaration of function ‘syz_open_pts’ [-Werror=implicit-function-declaration]
        		cc1: all warnings being treated as errors
        		
        		compiler invocation: gcc [-Wall -Werror -O1 -g -o /tmp/syz-executor613277355 -pthread -DGOOS_openbsd=1 -DGOARCH_amd64=1 -x c - -m64 -static]
```

This is related to generation of C reproducers and how they are combined from pieces of `executor/common_openbsd.h`.
You should be able to reproduce it locally by running:
`go test -short -run=TestGenerate/openbsd ./pkg/csource`
And when that passes also run it without -short flag.

dvyukov commented on pkg/report/openbsd.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func (ctx *openbsd) Parse(output []byte) *Report {
+	rep := &Report{
```
I think this whole function can now be replaced with `simpleLineParser`. See `pkg/report/gvisor.go` for an example.

mptre replied:
That worked, thanks!

dvyukov commented:
I don't see where you add support for OpenBSD in executor code.
You need to at least update this part of executor.cc
```
#if GOOS_linux
#include "executor_linux.h"
#elif GOOS_fuchsia
#include "executor_fuchsia.h"
#elif GOOS_akaros
#include "executor_akaros.h"
#elif GOOS_freebsd || GOOS_netbsd
#include "executor_bsd.h"
#elif GOOS_windows
#include "executor_windows.h"
#elif GOOS_test
#include "executor_test.h"
#else
#error "unknown OS"
#endif
```

Not sure if you need to touch anything in executor_bsd.h or in common_bsd.h.

dvyukov commented on pkg/host/host_openbsd.go:
```diff
@@ -0,0 +1,20 @@
+}
+
+func init() {
+	checkFeature[FeatureCoverage] = unconditionallyEnabled
```
This looks wrong, because we don't have coverage support in executor. I guess you copied it from netbsd, but I got it wrong after a refactoring. Now fixed in:
https://github.com/google/syzkaller/commit/bf3cf0c965acde4217280429327c9012b37aaf17

So just drop this whole init function.

dvyukov commented on sys/openbsd/init.go:
```diff
@@ -0,0 +1,80 @@
+
+func InitTarget(target *prog.Target) {
+	arch := &arch{
+		MAP_ANONYMOUS: target.ConstMap["MAP_ANON"],
```
You seem to not need the mmap consts because we use the helper.

dvyukov commented on sys/openbsd/init.go:
```diff
@@ -0,0 +1,80 @@
+	switch c.Meta.CallName {
+	case "mmap":
+		// Add MAP_FIXED flag, otherwise it produces non-deterministic results.
+		c.Args[3].(*prog.ConstArg).Val |= arch.MAP_FIXED
```
We have a common helper for this function that is used on linux/netbsd/freebsd. netbsd/freebsd have 0 custom code here now. Can you just use that function here? Not sure if you are doing something more complex with mknod here. If yes, could we improve the common function to do it too, and then use the common function?

mptre replied:
I did a compromise and started using the helper function but kept the `VBAD` parts since it removes unwanted crashes.

mptre commented on pkg/host/host_openbsd.go:
```diff
@@ -0,0 +1,20 @@
+	return true, ""
+}
+
+func EnableFaultInjection() error {
```
Fixed.

mptre commented:
- The tests should hopefully work by now. I cannot run the on OpenBSD since the cpp(1) implementation from Clang does not support `-fdirectives-only`. Didn't investigate that much time into fixing this...

- The executor parts are now also pushed.

- Regarding the syscall descriptions: I did copy them from NetBSD initially and removed everything related to Linux-compat, added `kqueue.txt` that hopefully could be shared between the BSDs, added two OpenBSD-specific syscalls `pledge(2)` and `unveil(2)`.

krytarowski commented:
On the ABI level BSD diverged too much to pretend a shared subset.

I recommend to keep BSD as distinct targets, also if something is shared between NetBSD and FreeBSD, it might be very different on other BSD and we will end up quickly with a lot of code such as '#if BSD && !OTHERBSD`.

dvyukov commented:
@krytarowski 
> On the ABI level BSD diverged too much to pretend a shared subset.
> I recommend to keep BSD as distinct targets

But there is a very large shared subset too, right. All of the fs, mm, socket, ipc calls, structures and flags. And it's not only the BSDs, also linux, darwin, fuchsia, akaros and potentially some other unix flavours. So we will have 7+ fold duplication there. Which is somewhat concerning. We put lots of effort into, say, linux, and it would be nice to immediately reuse that for an OS for which nobody has lots of resources (at least initially until we can prove that it is actually useful).

mptre commented:
The latest build failed on [Travis](https://travis-ci.org/google/syzkaller/jobs/421648882) for Go 1.10:

```
--- FAIL: TestGenerate (7.33s)
    --- FAIL: TestGenerate/linux/arm64 (0.21s)
    	csource_test.go:51: seed=1535474048909453807
        --- FAIL: TestGenerate/linux/arm64/0/39 (10.55s)
<snip>
collect2: error: ld terminated with signal 11 [Segmentation fault], core dumped
```

yikes...

dvyukov commented:
> The latest build failed on Travis for Go 1.10:

This is "OK". I guess it's an OOM, we periodically see these, don't know what to do yet. So if 1 of them passes, it generally means that the change is good.

I've skimmed through code and it looks good to me. I did not read every single line super carefully, because it's large. I am merging this now, we can fix up things later in smaller changes if necessary.

krytarowski commented:
> But there is a very large shared subset too, right. All of the fs, mm, socket, ipc calls, structures and flags.

Even if something might offer a similar feature it might contain different implementation details. We noted kqueue in this thread - a typical BSD interface, `struct kevent` differs on NetBSD and OpenBSD.

We were trying to do a similar thing in sanitizers and we ended up with `lib/sanitizer_common/sanitizer_platform_limits_{linux,netbsd,openbsd,solaris,posix}` as otherwise it was unmaintainable and every new kernel was difficult to handle.

krytarowski commented:
BSDs diverged in 1995 and after the years a lot of internals started to differ.  This situation will get only worse over time.

dvyukov commented:
> BSDs diverged in 1995 and after the years a lot of internals started to differ. This situation will get only worse over time.

Lots of that is POSIX.

> Even if something might offer a similar feature it might contain different implementation details. We noted kqueue in this thread - a typical BSD interface, struct kevent differs on NetBSD and OpenBSD.

I wonder what's the ratio between common interfaces and os-specific. Say, if kqueue is the same except for the kevent struct, then we can have a common kqueue description and then say that if an os includes this file then it needs to define own kevent struct elsewhere (all these *.txt files are in "global namespace" and order does not matter, so it's perfectly fine to just define kevent in any other .txt file).

For example, BSD sockets is quite large API, it would be nice to share common parts, but os-specific setsockopt's can go into os-specific files.

I don't know, maybe if we will try to share them, it will turn out that it's worse than what we have now. But I think we should try to prototype that sharing and see how bad/good it is actually in practice.

dvyukov commented:
OS-specific syscalls can go into OS-specific files.
For flags I was thinking about the following syntax:
```
// in a common file
mmap_flags = ... posix mmap flags...

// in a linux-specific file:
mmap_flags += ... linux flags ...
```

Re ptrace, we can decide for each subsystem separately. If ptrace is radically different, then we may not create a common file.

Copy-pasting is very fast solution in short term, but it creates problems long-term.
We had lots of copy-pasting in executor/* and that made adding new features hard, some important features were not implemented for anything other than linux, and everything other than linux were lower quality, and I periodically broken bsd's.
Recently I unified most of code and it all become better.
Specifically for descriptions, it's super easy to make bugs there, especially stupid bugs, and the descriptions are effectively untested. It makes it even more important to not have it copy-pasted, because eventually somebody will notice the stupid typo (today, most likely on linux) and fix it, but not necessary in 10 copies. We do have lots of such cases, I suspect it already affects bsd's.

krytarowski commented:
OK, so let's give it a try and evaluate in future.

blackgnezdo commented on executor/common_bsd.h:
```diff
@@ -18,3 +18,34 @@ static int do_sandbox_none(void)
+#include <termios.h>
+#include <util.h>
+#else
+// Needed when compiling on Linux.
```
Could somebody explain this to me? Does common_bsd.h need to compile cleanly on Linux? I am thinking of extending this with network packet injection. It might be harder to make cross-platform compatible.
